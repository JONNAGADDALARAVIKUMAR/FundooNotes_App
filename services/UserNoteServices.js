import Firebase from '../config/Firebase';
import KeyChain from 'react-native-keychain';

class UserNoteServices {

    getDetailsFromFirebase = () => {
        return new Promise(async (resolve, reject) => {
            const user = await KeyChain.getGenericPassword();
            const userDetails = JSON.parse(user.password);
            Firebase.database().ref('notes/' + userDetails.user.uid).once('value').then(async snapShot => { 
                resolve(snapShot.val())
            })
            .catch(error => reject(error))
        })
    }
    
    addNoteToFirebase = async (noteKey, title, note, Deletedstatus) => {
        return new Promise(async (resolve, reject) => {
            const user = await KeyChain.getGenericPassword();
            const userDetails = JSON.parse(user.password);
            const notes = {
                title : title,
                note : note,
                isDeleted: Deletedstatus
            }
            Firebase.database().ref('notes/' + userDetails.user.uid  + '/' + noteKey).set({
                notes : notes
            })
            .then(() => resolve('success'))
            .catch(error => reject(error))
            })
        }

        restoreNoteInFirebase = async (title, note, notekey) => {
            return new Promise(async (resolve, reject) => {
                const user = await KeyChain.getGenericPassword();
                const userDetails = JSON.parse(user.password);
                const notes = {
                    title : title,
                    note : note,
                    isDeleted : false
                }
                Firebase.database().ref('notes/' + userDetails.user.uid  + '/' + notekey).set({
                    notes : notes
                })
                .then(() => resolve('success'))
                .catch(error => reject(error))
            })
        }
    }

export default new UserNoteServices();
