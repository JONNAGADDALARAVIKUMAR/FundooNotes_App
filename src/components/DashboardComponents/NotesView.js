import React, { Component } from 'react';
import {View, Text, ImageBackground, Dimensions, ScrollView, Image} from 'react-native';
import {strings} from '../../Languages/strings';
import DashBoardScreenStyles from '../../styles/DashBoardScreenStyles';
import { Button, Card, Paragraph, Title } from 'react-native-paper';
import NoteViewStyles from '../../styles/NoteViewStyles';
import SQLiteStorageServices from '../../../services/SQLiteStorageServices';
import {storeUserID, 
        storeEditNotesDetails, 
        storeNoteKeyToUpdateNotes, 
        storeSelectedLabelKeys,
        storelabelsAndLabelKeys} from '../../redux/actions/CreateNewLabelAction';
import { connect } from 'react-redux';
import KeyChain from 'react-native-keychain';
import SQLiteLabelServices from '../../../services/SQLiteLabelServices';

class NotesView extends Component {
    constructor() {
        super();

        const isPortrait = () => {
            const dim = Dimensions.get('screen');
            return dim.height >= dim.width;
        };
    
        this.state = {
            changeLayout: false,
            height: Dimensions.get('window').height,
            width: Dimensions.get('window').width,
            orientation: isPortrait() ? 'portrait' : 'landscape',
            notes: [],
            isEmpty: true,
            labels: []
        };
    
        Dimensions.addEventListener('change', () => {
            this.setState({
                orientation: isPortrait() ? 'portrait' : 'landscape'
            });
        });
    }

    componentDidMount = async () => {

        const credential = await KeyChain.getGenericPassword();
        const UserCredential = JSON.parse(credential.password);
        await this.props.storeUserId(UserCredential.user.uid)

        await SQLiteStorageServices.getDetailsFromSQLiteDatabase()
        .then(async (results) => {
            var temp = []
            if(results.rows.length != 0) {
                for (let i = 0; i < results.rows.length; ++i)
                temp.push(results.rows.item(i));
                await this.setState({
                    notes : temp
                })
                temp.map(async (notes) => {
                    (notes.isDeleted == this.props.status)
                    ? this.setState({
                        isEmpty: false
                    })
                    : null
                })
            }
            SQLiteLabelServices.getLabelsFromSQliteStorage(this.props.userId)
            .then(results => {
                if(results.rows.length > 0) {
                    let labels = [];
                    for(let i = 0; i<results.rows.length; i++ ) {
                        labels.push(results.rows.item(i))
                    }
                    this.setState({
                        labels: labels
                    })
                    this.props.storelabelsAndLabelKeys(labels)
                }
            })
            .catch(error => console.log(error))
        })
    }

    handleDetailsToUpdateSQLite = (noteKey, Title, Notes, isDeleted, isArchived, Labels) => {
        const notes = {
            title: Title,
            note: Notes,
            labels: Labels,
            isArchived: isArchived,
            isDeleted: isDeleted
        }
        this.props.storeEditNotesDetails(notes)
        this.props.storeNoteKeyToUpdateNotes(noteKey)
        this.props.storeSelectedLabelKeys(JSON.parse(Labels))

        this.props.navigation.push('AddNewNotes')
    }

    handleDeletedNotesToUpdate = (noteKey, Title, Notes, isDeleted, isArchived, Labels) => {
        const notes = {
            title: Title,
            note: Notes,
            labels: Labels,
            isArchived: isArchived,
            isDeleted: isDeleted
        }
        this.props.storeEditNotesDetails(notes)
        this.props.storeNoteKeyToUpdateNotes(noteKey)
        this.props.storeSelectedLabelKeys(JSON.parse(Labels))
        
        this.props.navigation.push('DeletedNoteView')
    }

    render() {
        return (
            <ImageBackground source = {require('../../assets/backgroundIcon.png')}
                style = {(this.state.orientation == 'portrait') 
                    ? {height: 510, width: 350, alignSelf: 'center'} 
                    : {height: 750, width: 530, alignSelf: 'center'}}>

                <ScrollView>
                    <View style = {NoteViewStyles.list_Style}>
                        {this.state.notes.length > 0 ?
                        this.state.notes.map(val => (
                            <React.Fragment key = {val.NoteKey}>
                                {val.isDeleted == this.props.status ? (
                                    <Card
                                        style = {this.props.changeLayout ? NoteViewStyles.list_grid_Container: NoteViewStyles.list_Container}
                                        onPress = {() => {
                                            (val.isDeleted) 
                                            ? this.handleDeletedNotesToUpdate(val.NoteKey, val.Title, val.Notes, val.isDeleted, val.isArchived, val.Labels) 
                                            : this.handleDetailsToUpdateSQLite(val.NoteKey, val.Title, val.Notes, val.isDeleted, val.isArchived, val.Labels)}
                                        }>
                                        <Card.Content>
                                            <Title>
                                                {val.Title}
                                            </Title>
                                            <Paragraph>
                                                {val.Notes}
                                            </Paragraph>
                                                <View style = {{flexWrap: 'wrap', flexDirection: 'row'}}>
                                                {(val.Labels.length > 0) ?
                                                    this.state.labels.map(labels => (
                                                        val.Labels.includes(labels.lebelKey) ?
                                                            <React.Fragment key = {labels.lebelKey}>
                                                                <View style = {{}}>
                                                                    <Text>{labels.labelName}</Text>
                                                                </View>
                                                            </React.Fragment>
                                                        :
                                                        null
                                                    ))
                                                :
                                                null}
                                                </View>
                                        </Card.Content>  
                                    </Card>)
                                : null}
                            </React.Fragment>
                        ))
                        : null}
                    </View>  

                    {(this.state.isEmpty) ? (
                    <View>
                        <Image style = {DashBoardScreenStyles.bulb_Style} source = {require('../../assets/bulb.png')}/>
                        <Text style = {DashBoardScreenStyles.Appear_Text_Style}>{strings.YourNoteswillApperHere}</Text>
                    </View>) : null }
                </ScrollView>
            </ImageBackground>
        )
    }
}

const mapStateToProps = state => {
    return {
        userId : state.createLabelReducer.userId,
        labelContent : state.createLabelReducer.labelContent,
    }
}

const mapDispatchToProps = dispatch => {
    return {
        storeUserId : (userId) => dispatch(storeUserID(userId)),
        storeEditNotesDetails : (notes) => dispatch(storeEditNotesDetails(notes)),
        storeNoteKeyToUpdateNotes : (noteKeyToUpdateNotes) => dispatch(storeNoteKeyToUpdateNotes(noteKeyToUpdateNotes)),
        storeSelectedLabelKeys : (selectedLabelKeys) => dispatch(storeSelectedLabelKeys(selectedLabelKeys)),
        storelabelsAndLabelKeys : (labelsAndLabelKeys) => dispatch(storelabelsAndLabelKeys(labelsAndLabelKeys)),
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(NotesView)