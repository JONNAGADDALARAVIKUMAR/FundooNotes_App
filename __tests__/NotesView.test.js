import React from 'react';
import {shallow} from 'enzyme';
import NotesView from '../src/components/DashboardComponents/NotesView';
import UserNoteServices from '../services/UserNoteServices';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const initialState = {
    userId : '',
    labelContent: [],
    labelNoteKeys: [],
    labels: [],
    showDailog: false,
    deleteLabelKey: null,
    selectedLabelKeys: [],
    notesArchived: false,
    editNotesDetails: null,
    noteKeyToUpdateNotes: '',
    labelsAndLabelKeys: [],
    labelAndKey: {},
    labelScreen: null
}
const store = mockStore(initialState)

describe('test NotesView Class', () => {
    it('test when render should match to snapshot', async () => {
        const component = shallow(<NotesView store = {store}/>)
        expect(component).toMatchSnapshot();
    })
    it('test component did mount method should get user details and Update State', async () => {
        const onDismissEvent = jest.fn();
        const component = shallow(<NotesView store = {store}/>)
        const instance = component.instance();
        UserNoteServices.getDetailsFromFirebase()
        .then((user) => {
            let notes = user ? user : {}
            expect(instance.state.notes).toBe(notes)
            expect(instance.state.isEmpty).toBe(false)
        })
    })
    it('test handleDetailsToUpdateSQLite method on press notes should navigate to AddNewNotes screen', async () => {
        const onPressEvent = jest.fn();
        const navigation = { navigation : jest.fn() }
        const component = shallow(<NotesView onPress = {onPressEvent} navigation = {navigation} store = {store}/>)
        const instance = component.instance();
        instance.handleDetailsToUpdateSQLite('123', 'Ravi', false)
        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.push).toHaveBeenCalledWith('AddNewNotes')
    })
    it('test handleDeletedNotesToUpdate method on press notes should navigate to DeletedNoteView screen', async () => {
        const onPressEvent = jest.fn();
        const navigation = { navigation : jest.fn() }
        const component = shallow(<NotesView store = {store} onPress = {onPressEvent} navigation = {navigation}/>)
        const instance = component.instance();
        instance.handleDeletedNotesToUpdate('123', 'Ravi', false)
        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.push).toHaveBeenCalledWith('DeletedNoteView')
    })
})