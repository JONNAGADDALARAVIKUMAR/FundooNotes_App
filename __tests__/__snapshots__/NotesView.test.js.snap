// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test NotesView Class test when render should match to snapshot 1`] = `ShallowWrapper {}`;
